File: CoinChange01.java
class CoinChange01 {
    public int coinChange(int[] coins, int amount) {
        if(coins.length==0 || coins==null) return -1;
         int[][] dp=new int[coins.length+1][amount+1];
        for (int j=1;j<amount+1;j++){
            dp[0][j]=amount+1;
        }
        for (int i=0;i<coins.length+1;i++){
            dp[i][0]=0;
        }
        for(int i=1;i<=coins.length;i++){
            for(int j=1;j<=amount;j++){
                if(coins[i-1]>j){
                    dp[i][j]=dp[i-1][j];
                }
                else{
                    dp[i][j]=Math.min(dp[i-1][j],(dp[i][j-coins[i-1]])+1);
                }
            }
        }
        int result= dp[coins.length][amount];
        if(result>amount) return -1;
        
        return result;
    }
}

File: CoinChange_01.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int m = coins.size();
        int n = amount;
     
        vector<vector<int>> dp(m+1, vector<int>(n+1));
        dp[0][0]=0;
        for(int j=1;j<=n;j++){
            dp[0][j] = INT_MAX-2;
        }
        for(int i=1;i<=m;i++){
            for(int j=1;j<=n;j++){
                if(j<coins[i-1]){
                    dp[i][j] = dp[i-1][j];
                }
                else{
                    dp[i][j] = min(dp[i-1][j], 1 + dp[i][j-coins[i-1]]);
                }
            }
        }
        if(dp[m][n]>= INT_MAX-2)return -1;
        else return dp[m][n];
    }
};
